platform/linux-generic/odp_packet.c

```diff
@@ -354,6 +354,15 @@ static inline void link_segments(odp_packet_hdr_t *pkt_hdr[], int num)
 			hdr->buf_hdr.seg[i].hdr  = buf_hdr;
 			hdr->buf_hdr.seg[i].data = buf_hdr->base_data;
 			hdr->buf_hdr.seg[i].len  = BASE_LEN;
+
+			/* Ensure that ref_cnt is set to 1 for debug builds.
+			 * This makes use of the comma operator to force
+			 * the store only for debug builds since the expression
+			 * (void_func(), 1) always evaluates to 1 so the
+			 * ODP_ASSERT always succeeds.
+			 */
+			ODP_ASSERT((odp_atomic_store_u32(&hdr->buf_hdr.ref_cnt,
+							 1), 1));
```

Comment:
Maybe it's better not to change ref_cnt book keeping for debug build. Change of behavior may hide the problem being debugged, etc.

Reply:
If we want to catch erroneous duplicate packet free calls I don't see how we can avoid that. That's a main source of application errors, especially when references are involved.

The lazy-increment shortcut is a clever efficiency trick for normal processing, but it's counter to what we want in a debug build.

Reply:
I'd expect that a debug build adds checks and printf, but does not change e.g. how an algorithm works.

Also store is not enough for double frees, you'd need to inc/dec and check that the value never gets <0.

Reply:
@psavol The double free guard is the `ODP_ASSERT(ref_cnt > 0)`checks  in `odp_packet_free()` and `odp_packet_free_multi()`. Incrementing and decrementing is already handled by the mainline code. This isn't changing how reference counts are used as the mainline code works the same. It's only removing the lazy increment optimization.